### PR TITLE
Set up External Portal Application Authentication with Azure AD OIDC

### PR DESCRIPTION
**Pull request by tlblessing**
See https://konghq.atlassian.net/browse/DOCS-1294. @aaronhmiller has provided the content in this google doc https://docs.google.com/document/d/1xXES9d-kzyLnj5U7oEbLXhy_70upuH7KhY4dJuH8ZBQ/edit#

Direct review link:

https://deploy-preview-2300--kongdocs.netlify.app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config/

Update: This ticket was logged by @skibum https://konghq.atlassian.net/browse/DOCS-1237, and Reza has requested a more generic documentation approach to 3rd-party IdPs for app reg auth. cc: @Jlawlzz

### CODE REVIEW

**Review by aaronhmiller — APPROVED** [review]
I've reviewed and these will be helpful to all deploying Azure AD for OIDC portal application authentication.

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+   secure location. You can only view the secret once.
+
+1. Under **Manifest**, update `accessTokenAcceptedVersion=2` (default is null).
+   The JSON for your application should look similar to this example:
+
+  ![Azure Manifest](/assets/images/docs/dev-portal/azure-manifest.png)
+
+## Create a Service and a Route in Kong
+
+1. Create a Service example:
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:32]:
```suggestion
1. Create a Service. For example:
```

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+   ```bash
+   http put :8001/services/httpbin-service-azure url=https://httpbin.org/anything
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:35]:
```suggestion
   $ http PUT :8001/services/httpbin-service-azure \
      url=https://httpbin.org/anything
```
Capitalize the method for consistency.

You can format any bash command example with `\` to denote arbitrary line breaks for readability. They won't change anything in the command itself, as long as you break in the middle of a name or a URL, for example.

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+1. Under **Manifest**, update `accessTokenAcceptedVersion=2` (default is null).
+   The JSON for your application should look similar to this example:
+
+  ![Azure Manifest](/assets/images/docs/dev-portal/azure-manifest.png)
+
+## Create a Service and a Route in Kong
+
+1. Create a Service example:
+
+   ```bash
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:34]: Most of the examples in our docs are in cURL, or in both cURL and HTTPie. We don't normally do HTTPie only. Can you turn these examples into cURL and HTTPie tabs, at the section level (right under the header)?

For example, this command would be the following in cURL:
```
curl -i -X PUT http://<admin-server>:8001/services/httpbin-service-azure2 \
   --data 'url=https://httpbin.org/'
```
You don't absolutely need `-i`, but it makes the output easier to read.

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+   http put :8001/services/httpbin-service-azure url=https://httpbin.org/anything
+   ```
+1. Create a Route example:
+
+   ```bash
+   http -f put :8001/services/httpbin-service-azure/routes/httpbin-route-azure paths=/httpbin-azure
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:40]:
```suggestion
   $ http -f PUT :8001/services/httpbin-service-azure/routes/httpbin-route-azure \
      paths=/httpbin-azure
```

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  ![Azure Manifest](/assets/images/docs/dev-portal/azure-manifest.png)
+
+## Create a Service and a Route in Kong
+
+1. Create a Service example:
+
+   ```bash
+   http put :8001/services/httpbin-service-azure url=https://httpbin.org/anything
+   ```
+1. Create a Route example:
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:37]:
```suggestion
1. Create a Route. For example:
```

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+## Map the OIDC and Application Registration Plugins to the Service
+
+Map the OpenID Connect and Application Registration plugins to the **Service**.
+The plugins must be applied to a Service to work properly.
+
+1. Configure the OIDC plugin for the Service:
+
+   ```bash
+   http -f :8001/services/httpbin-service-azure/plugins name=openid-connect config.issuer=https://login.microsoftonline.com/<your_tenant_id>/v2.0 config.display_errors=true config.client_id=<your_client_id> config.client_secret="<your_client_secret>" config.consumer_claim=aud config.scopes=openid config.scopes=<your_client_id>/.default
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:51]:
```suggestion
  $ http -f :8001/services/httpbin-service-azure/plugins \
     name=openid-connect \
     config.issuer=https://login.microsoftonline.com/<your_tenant_id>/v2.0 \
     config.display_errors=true \
     config.client_id=<your_client_id> \
     config.client_secret="<your_client_secret>" \
     config.consumer_claim=aud \
     config.scopes=openid \
     config.scopes=<your_client_id>/.default
```

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+   ```
+
+1. Configure the Application Registration plugin for the Service:
+
+   ```bash
+   http -f :8001/services/httpbin-service-azure/plugins name=application-registration config.auto_approve=true config.display_name="For Azure" config.description="Uses consumer claim with various values (sub, aud, etc.) as registration id to support different flows and use cases." config.show_issuer=true
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:57]:
```suggestion
   $ http -f :8001/services/httpbin-service-azure/plugins \
      name=application-registration \
      config.auto_approve=true \
      config.display_name="For Azure" \
      config.description="Uses consumer claim with various values (sub, aud, etc.) as registration id to support different flows and use cases." \
      config.show_issuer=true
```

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+   ```
+
+1. Get an access token using the Client Credential workflow and convert the token
+   into a JSON Web Token (JWT). You can get the token from Azure or Kong. Replace
+   the placeholder values for `<your_tenant_id>`, `<your_client_id>`, and `<your_client_secret>`.
+
+   - Get a token from Azure:
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:64]: We currently have a bug in the output that reads unordered lists as implicit entries in an ordered list, so the numbering is broken. Change this entry and the following one to either a definition list or a numbered list (see https://deploy-preview-2300--kongdocs.netlify.app/hub/plugins/overview/#terminology for example formatting for a definition list).

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+     ```bash
+     https -f POST "https://login.microsoftonline.com/<your_tenant_id>/oauth2/v2.0/token" scope=<your_client_id>/.default grant_type=client_credentials -a <your_client_id>:<your_client_secret>
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:67]:
```suggestion
     $ https -f POST "https://login.microsoftonline.com/<your_tenant_id>/oauth2/v2.0/token" \
        scope=<your_client_id>/.default \
        grant_type=client_credentials \
        -a <your_client_id>:<your_client_secret>
```

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+     ```   
+
+   - Get a token from Kong:
+
+     ```bash
+     https -f POST "https://localhost:8443/httpbin-azure/oauth2/v2.0/token" grant_type=client_credentials -a <your_client_id>:<your_client_secret> --verify NO
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:73]:
```suggestion
     $ https -f POST "https://localhost:8443/httpbin-azure/oauth2/v2.0/token" \
        grant_type=client_credentials \
        -a <your_client_id>:<your_client_secret> \
        --verify NO
```
Should `localhost` really be used here? Can we replace it with <admin-hostname>? Same goes for all other examples using localhost in this topic.

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+1. Paste the access token obtained from the previous step into
+   [JWT](https://jwt.io), and copy the value for the
+   [aud (audience)](https://tools.ietf.org/html/rfc7519#section-4.1.3) claim to
+   your clipboard. You will use the `aud` value as your **Reference ID** for the next step.
+
+## Create an Application in Kong
+
+1. Log in to your Developer Portal and create a new application:
+   - Select the **My Apps** menu -> **New Application**.
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:84]: Same issue as a previous comment; turn this list into a numbered list.

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+Follow these instructions to test your client credentials or authorization code
+flows with your Azure AD implementation.
+
+### Test Client Credentials Flow
+
+1. Get a token:
+
+   ```bash
+   https -f POST "https://localhost:8443/httpbin-azure/oauth2/v2.0/token" grant_type=client_credentials -a <your_client_id>:<your_client_secret> --verify NO
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:118]:
```suggestion
   $ https -f POST "https://localhost:8443/httpbin-azure/oauth2/v2.0/token" \
      grant_type=client_credentials \
      -a <your_client_id>:<your_client_secret> \
      --verify NO
```

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+   ```
+
+2. Use the token in an authorization header to retrieve the data:
+
+   ```bash
+   http :8000/httpbin-azure Authorization:'bearer <token_from_above>'
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:124]:
```suggestion
   $ http :8000/httpbin-azure Authorization:'bearer <token_from_above>'
```

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+
+   Replace `<token_from_above>` with the bearer token you generated in the previous step.
+
+### Test Authorization Code Flow
+
+In your browser, go to `http://localhost:8000/httpbin-azure`. Replace `localhost`
+with your host information.
+
+You should be guided through a log in process within Azure and then the results
+delivered in your browser.
```
Comment (lena-larionova) [inline app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:135]:
```suggestion
If authorization setup was successful, you should be guided through a log in process within Azure.
```
I don't think you need the last part of the sentence here, but that's up to you.

Comment (tlblessing) [reply app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:135]: stet on Aaron's content

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md (hunk 8 above)
Comment (tlblessing) [reply app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:79]: > We currently have a bug in the output that reads unordered lists as implicit entries in an ordered list, so the numbering is broken. Change this entry and the following one to either a definition list or a numbered list (see https://deploy-preview-2300--kongdocs.netlify.app/hub/plugins/overview/#terminology for example formatting for a definition list).

@lena-larionova I didn't even notice this at first. I just fixed this issue in another ticket for the dev portal. So this is an either or situation, so ul was perfect, numbered list is not appropriate here. Will this bug be fixed in the future? I could just use indents in the interim, or defn list.

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md (hunk 11 above)
Comment (tlblessing) [reply app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md]: Done in commit 772e294

File: app/_data/docs_nav_ee_2.1.x.yml [hunk 15]
```diff
@@ -424,6 +424,8 @@
               url: /developer-portal/administration/application-registration/enable-application-registration
             - text: Set up Okta and Kong for external OAuth
               url: /developer-portal/administration/application-registration/okta-config
+            - text: Set up Azure AD and Kong for external authentication
```
[the remark this answers is not in the document]
Comment (tlblessing) [reply app/_data/docs_nav_ee_2.1.x.yml]: `Should also adjust the topic entry above this one - I know it's not part of this PR, so it's up to you whether you want to do it here or separately.`

I have a PR in progress for that topic; will do it there.

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md (hunk 10 above)
Comment (tlblessing) [reply app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md]: Done in commit ae436d7

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md (hunk 3 above)
Comment (tlblessing) [reply app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:34]: @lena-larionova yes we can show cURL as well. I had mentioned that to @aaronhmiller. The OIDC example uses Kong Manager GUI and doesn't show any commands. So we need to be consistent (something for me to keep in mind for the 3rd party/OIDC updates PR). I'm not sure if we'll get to publish/keep these IdP-specific examples.

I don't have anything set up in azure. I'll need someone to look over my guesses at converting httpie -> curl.

Comment (tlblessing) [reply app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md:34]: I think your example above has an extraneous 2 in `azure2`

File: app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md (hunk 8 above)
Comment (lena-larionova) [reply app/enterprise/2.1.x/developer-portal/administration/application-registration/azure-oidc-config.md]: Yes, it'll be fixed in the future, once we either get Vlad to do it or I figure it out. Indents or definition list will work in the interim.

**Comment by tlblessing** [comment]
@aaronhmiller can you please review and verify the cURL examples with your setup?

**Review by aaronhmiller — APPROVED** [review]
I reviewed the cURL commands and tested them. My input has been taken into account and for me, this is ready for publishing.

**Review by lena-larionova — APPROVED** [review]
LGTM!

**Comment by tlblessing** [comment]
> LGTM!

Thank you @lena-larionova!

**Pull request merged**